[quant] fill_ path for quantized tensors

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43304 [quant] quantized path for ConstantPadNd
* **#43303 [quant] fill_ path for quantized tensors**

Differential Revision: [D23231947](https://our.internmc.facebook.com/intern/diff/D23231947)

Commit:
[quant] fill_ path for quantized tensors

[ghstack-poisoned]

ailzhang reviewed (commented):
ehhh what's the main goal of this PR (perf?) could you please add a bit context?

z-a-f commented:
> ehhh what's the main goal of this PR (perf?) could you please add a bit context?

It's introducing the `fill_` for quantized tensors, which is needed for different ops (specifically for the ConstantPad)

Commit:
Update on "[quant] fill_ path for quantized tensors"


Differential Revision: [D23231947](https://our.internmc.facebook.com/intern/diff/D23231947)

[ghstack-poisoned]

Commit:
Update on "[quant] fill_ path for quantized tensors"


Differential Revision: [D23231947](https://our.internmc.facebook.com/intern/diff/D23231947)

[ghstack-poisoned]

Commit:
Update on "[quant] fill_ path for quantized tensors"


Differential Revision: [D23231947](https://our.internmc.facebook.com/intern/diff/D23231947)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantized_tensor.py:
```diff
@@ -484,6 +484,32 @@ def test_qtensor_clone(self):
+                q = torch._empty_affine_quantized([numel], scale=scale,
+                                                  zero_point=zero_point,
+                                                  device=device, dtype=qtype)
+                q_filled = q.fill_(-1)
```
NIT: Should this be fill_with?

z-a-f replied:
My bad -- will change before landing!

raghuramank100 commented on test/quantization/test_quantized_tensor.py:
```diff
@@ -484,6 +484,32 @@ def test_qtensor_clone(self):
+                     [torch.int8, torch.uint8, torch.int32])
+
+        fill_with = -1
+        int_repr = int(round(fill_with / scale + zero_point))
```
What happens when this goes out of range for each of the dtypes?

raghuramank100 commented on aten/src/ATen/native/Fill.cpp:
```diff
@@ -19,6 +19,11 @@ namespace {
 Tensor& fill_out(Tensor& self, Scalar value) {
+  if (self.is_quantized()) {
+    at::Tensor out = at::ones(self.sizes()).to(kFloat) * value;
+    self.copy_(out);
```
How are out of range fill values handled?

z-a-f replied:
The `copy_` method already handles the logic of quantizing and taking care of the boundaries

z-a-f replied:
I will put an inline comment about that in here

Commit:
Update on "[quant] fill_ path for quantized tensors"


Differential Revision: [D23231947](https://our.internmc.facebook.com/intern/diff/D23231947)

[ghstack-poisoned]

Commit:
Update on "[quant] fill_ path for quantized tensors"


Differential Revision: [D23231947](https://our.internmc.facebook.com/intern/diff/D23231947)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/Fill.cpp:
```diff
@@ -19,6 +19,13 @@ namespace {
 Tensor& fill_out(Tensor& self, Scalar value) {
+  if (self.is_quantized()) {
+    at::Tensor out = at::ones(self.sizes()).to(kFloat) * value;
+    out = out.to(self.device());
```
is this needed?  Can we assert that `out.device() == self.device()` instead?

z-a-f replied:
We cannot assert it as the out created in this block. If we don't assign the device, we will have to trust the copy_ to move the tensors across the devices, which is an antipattern :/

Commit:
Update on "[quant] fill_ path for quantized tensors"


Differential Revision: [D23231947](https://our.internmc.facebook.com/intern/diff/D23231947)

[ghstack-poisoned]

Commit:
Update on "[quant] fill_ path for quantized tensors"


Differential Revision: [D23231947](https://our.internmc.facebook.com/intern/diff/D23231947)

[ghstack-poisoned]

Commit:
Update on "[quant] fill_ path for quantized tensors"


Differential Revision: [D23231947](https://our.internmc.facebook.com/intern/diff/D23231947)

[ghstack-poisoned]
